php-checks/src/main/java/org/sonar/php/checks/security/FileSystemUsageCheck.java

```diff
@@ -0,0 +1,134 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.php.checks.utils.CheckUtils;
+import org.sonar.plugins.php.api.tree.SeparatedList;
+import org.sonar.plugins.php.api.tree.Tree.Kind;
+import org.sonar.plugins.php.api.tree.declaration.NamespaceNameTree;
+import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
+import org.sonar.plugins.php.api.tree.expression.LiteralTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+
+@Rule(key = "S4797")
+public class FileSystemUsageCheck extends PHPVisitorCheck {
+
+  private static final String MESSAGE = "Make sure this file handling is safe here.";
+
+  private static final Map<String, FunctionArgsMatcher> FILE_SYSTEM_FUNCTIONS = buildFileSystemFunctions();
+
+  private static Map<String, FunctionArgsMatcher> buildFileSystemFunctions() {
+    HashMap<String, FunctionArgsMatcher> map = new HashMap<>();
+    map.put("chgrp", new FunctionArgsMatcher(2, 2, false, 0));
+    map.put("chmod", new FunctionArgsMatcher(2, 2, false, 0));
+    map.put("chown", new FunctionArgsMatcher(2, 2, false, 0));
+    map.put("copy", new FunctionArgsMatcher(2, 2, true, 0, 1));
+    map.put("delete", new FunctionArgsMatcher(1, 1, false, 0));
+    map.put("file", new FunctionArgsMatcher(1, 2, true, 0));
+    map.put("file_get_contents", new FunctionArgsMatcher(1, 2, true, 0));
+    map.put("file_put_contents", new FunctionArgsMatcher(2, 3, true, 0));
+    map.put("fopen", new FunctionArgsMatcher(2, 3, true, 0));
+    map.put("lchgrp", new FunctionArgsMatcher(2, 2, false, 0));
+    map.put("lchown", new FunctionArgsMatcher(2, 2, false, 0));
+    map.put("move_uploaded_file", new FunctionArgsMatcher(2, 2, false, 0, 1));
+    map.put("parse_ini_file", new FunctionArgsMatcher(1, 3, true, 0));
+    map.put("readfile", new FunctionArgsMatcher(1, 2, true, 0));
+    map.put("rmdir", new FunctionArgsMatcher(1, 1, false, 0));
+    map.put("tmpfile", new FunctionArgsMatcher(0, 0, true));
+    map.put("unlink", new FunctionArgsMatcher(1, 1, false, 0));
+    return map;
+  }
+
+  @Override
+  public void visitFunctionCall(FunctionCallTree tree) {
+    if (tree.callee().is(Kind.NAMESPACE_NAME)) {
+      String qualifiedName = ((NamespaceNameTree) tree.callee()).qualifiedName();
+      FunctionArgsMatcher argsMatcher = FILE_SYSTEM_FUNCTIONS.get(qualifiedName);
+      if (argsMatcher != null && argsMatcher.matches(tree)) {
+        context().newIssue(this, tree, MESSAGE);
+      }
+    }
+    super.visitFunctionCall(tree);
+  }
+
+  private static class FunctionArgsMatcher {
+
+    private static final String FILE_SCHEME = "file";
+
+    private static final String SCHEME_SEPARATOR = "://";
+
+    private static final Set<String> WRAPPER_SCHEMES = new HashSet<>(Arrays.asList(
+      "rar", "zlib", "bzip2", "zip", "compress.zlib", "compress.bzip2", "glob", "ogg"));
```

Comment:
Since we want to distinguish between local filesystem and network accesses, it might be easier to use a whitelist (for `http`, `https`, `ftp`, and `ssh*`) instead of a blacklist to validate the hardcoded strings.

Reply:
I replaced the logic with a whitelist of network scheme